ruby_02-web_applications_with_ruby/outlines/rack_test_in_sinatra.markdown

```diff
@@ -9,200 +9,66 @@ tags: testing, tdd, sinatra, http, rack
-
-```ruby
-class FilmFile < Sinatra::Base
-  get '/genres' do
-    @genres = Genre.all
-    erb :genres_index
-  end
-
-  post '/genres' do
-    Genre.create(params[:genre])
-    status 200
-    body "Genre created."
-  end
-end
-```
-
-Cool, it works. But what if someone tries to create a task without a title? We need to [validate](http://guides.rubyonrails.org/active_record_validations.html) our data. Let's write a test first:
-
-```ruby
-  def test_cannot_create_a_genre_without_a_name
-    post '/genres', { genre: { } }
-    assert_equal 0, Genre.count
-    assert_equal 400, last_response.status
-    assert_equal "missing name", last_response.body
-  end
-```
-
-Normally at this point, we would drop down into a model test and write a test for our validations. For the purpose of this lesson, we'll skip that for now. 
-
-In our Genre model:
-
-```ruby
-class Genre < ActiveRecord::Base
-  validates_presence_of :name
-end
-```
-
-In our controller:
-
-```ruby
-  post '/genres' do
-    genre = Genre.new(params[:genre])
-    if genre.save
-      status 200
-      body "Genre created."
-    else
-      status 400
-      body "missing name"
-    end
-  end
-```
-
-What happens to the `genre` object if `if genre.save` doesn't get hit? You can access the following things:
-
-```ruby
-genre.errors.full_messages
-genre.errors[:title]
-```
-
-Let's try it out in tux. 
-
-## Key Points and Other Things
-
-* Use the `rack-test` gem to test rack applications (e.g. Sinatra, Rails)
-* Rack-test hooks in at the level of Rack, so it calls your code the same as a real web request, but with mock objects
-* Declare an `app` method so it knows what Rack app to use
-
-* What can you do with Rack::Test? 
-  * get access to the methods by including `include Rack::Test::Methods`
-  * make a request: (`get/post/put/patch/delete`)
-  * pass params by providing a params hash as the second argument: `post '/', { title: "My Idea", description: "it's hard to come up with descriptions" }`
-  * follow a redirect: `follow_redirect!`
-  * get the request or response: `last_request` or `last_response`
-
-* What to use `last_response` for:
-  * `last_response.status` codes: (200, 404, 302, 500, etc) -- see all status codes [here](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)
-  * `last_response.body`
-
-* Make assertions about the body
-  * look for relevant strings using normal string methods like `.include?`, etc.
-
-## Resources
-
-* [Testing Sinatra with Rack::Test](http://www.sinatrarb.com/testing.html)
-* [Rack Test Docs](https://github.com/brynary/rack-test/)
+## Setup (before lesson)
+
+* Clone the repo [here](https://github.com/s-espinosa/rack_test_in_sinatra)
+
+## Review: Unit and Feature Tests (Slides - 5 mins)
+
+* Unit tests - our specific function works the way we expect with given inputs
+    * Action: method calls on an object
+    * Result: return values/object state
+    * Test: example test
```

Comment:
What are these example tests? Are these on your slides? If so, it might be useful to also put them in here!